File: ex_src/book_shop.ex
defmodule BookShop do

  # Types

  @type cat :: {:cat, binary}
  @type address :: {:address, binary}

  defmodule Book do
    defstruct [:id, :title, :author]
    @enforce_keys [:id, :title, :author]
    @type t :: %Book{id: binary, title: binary, author: binary}
  end

  defmodule Order do
    alias BookShop, as: BS
    defstruct [:customer, :shipping_address, :books]
    @enforce_keys [:customer, :shipping_address, :books]
    @type t :: %Order{customer: BS.cat, shipping_address: BS.address, books: [Book.t]}
  end


  # Test Data

  @spec test_data() :: map
  def test_data do
    %{
      "cat" => "Tihon",
      "address" => "Coolcat str 7/42 Minsk Belarus",
      "books" => [
        %{
          "title" => "Domain Modeling Made Functional",
          "author" => "Scott Wlaschin"
        },
        %{
          "title" => "Удовольствие от Х",
          "author" => "Стивен Строгац"
        },
        %{
          "title" => "Distributed systems for fun and profit",
          "author" => "Mikito Takada"
        }
      ]
    }
  end


  # Module API

  @spec validate_incoming_data(map) :: {:ok, map} | {:error, :invalid_incoming_data}
  def validate_incoming_data json_data do
    case rand_success() do
      true -> {:ok, json_data}
      false -> {:error, :invalid_incoming_data}
    end
  end


  @spec validate_cat(binary) :: {:ok, cat} | {:error, :cat_not_found}
  def validate_cat cat_name do
    case rand_success() do
      true -> {:ok, {:cat, cat_name}}
      false -> {:error, :cat_not_found}
    end
  end


  # Internal functions

  @spec rand_success() :: boolean
  def rand_success do
    rand = :rand.uniform 10
    rand > 1
  end

end

File: src_hs/BookShop.hs
module BookShop where

import qualified System.Random as SR

data Cat = Cat String deriving Show

data Address = Address String deriving Show

data Book =
  Book { id :: String
       , title :: String
       , author :: String
       } deriving Show

data Order =
  Order { customer :: Cat
        , shipping_address :: Address
        , books :: [Book]
        } deriving Show

data ValidationError
  = InvalidIncomingData
  | CatNotFound
  | InvalidAddress
  | BookNotFound
  deriving Show


type JsonData = (String, String, [(String, String)])

test_data :: JsonData
test_data =
    ("Tihon", "Coolcat str 7/42 Minsk Belarus",
    [ ("Scott Wlaschin", "Domain Modeling Made Functional")
    , ("Стивен Строгац", "Удовольствие от Х")
    , ("Mikito Takada", "Distributed systems for fun and profit")
    ])


test_data' :: JsonData
test_data' =
    ("Marfa", "Coolcat str 7/42 Minsk Belarus",
    [ ("Scott Wlaschin", "Domain Modeling Made Functional")
    , ("Mikito Takada", "Distributed systems for fun and profit")
    ])


validate_incoming_data :: JsonData -> Either ValidationError JsonData
validate_incoming_data json_data =
  Right json_data


validate_cat :: String -> Either ValidationError Cat
validate_cat cat_name =
  case cat_name of
    "Tihon" -> Right $ Cat cat_name
    _ -> Left CatNotFound


validate_address :: String -> Either ValidationError Address
validate_address addr_str =
  Right $ Address addr_str


get_book :: String -> String -> Either ValidationError Book
get_book title author =
  Right $ Book  "ISBN 978-5-00057-917-6" title author


create_order :: Cat -> Address -> [Book] -> Order
create_order cat address books =
  Order cat address books
